MobileWallet/Backup/ICloudBackup.swift

```diff
@@ -0,0 +1,342 @@
+        }
+    }
+
+    private func downloadBackup(walletFolder: String, completion:((_ url: URL) throws -> Void)) throws {
+        let fileManager = FileManager.default
+        let icloudFolderURL = try iCloudDirectory().appendingPathComponent(walletFolder)
+        let urls = try fileManager.contentsOfDirectory(atURL: icloudFolderURL, sortedBy: .created, options: [])
+
+        if let backupUrl = urls.last {
+            var lastPathComponent = backupUrl.lastPathComponent
+            let folderPath = backupUrl.deletingLastPathComponent().path
+            // if the last path component contains the “.icloud” extension. If yes the file is not on the device else the file is already downloaded.
+
+            if lastPathComponent.contains(".icloud") {
+                lastPathComponent.removeFirst()
+                let downloadedFilePath = folderPath + "/" + lastPathComponent.replacingOccurrences(of: ".icloud", with: "")
+                var isDownloaded = false
+                try fileManager.startDownloadingUbiquitousItem(at: backupUrl)
+
+                while !isDownloaded {
+                    if fileManager.fileExists(atPath: downloadedFilePath) {
+                        isDownloaded = true
+                        try completion(URL(fileURLWithPath: downloadedFilePath))
+                    }
+                }
+            } else {
+                try completion(backupUrl)
+            }
+        }
+    }
+
+    private func existsWallets() throws -> [String] {
+        var wallets = [String]()
+        let icloudFolderURL = try iCloudDirectory()
+        do {
+            let urls = try FileManager.default.contentsOfDirectory(at: icloudFolderURL, includingPropertiesForKeys: nil, options: [])
+            urls.forEach({
+                if $0.lastPathComponent != "Documents" {
+                    wallets.append($0.lastPathComponent)
+                }
+            })
+            if wallets.isEmpty {
+                throw ICloudBackupWalletError.noAnyBackups
+            }
+            return wallets
+        } catch {
+            throw ICloudBackupWalletError.noAnyBackups
+        }
+    }
+
+    private func addNotificationObservers() {
+        NotificationCenter.default.addObserver(forName: NSNotification.Name.NSMetadataQueryDidStartGathering, object: query, queue: query.operationQueue) { [weak self] (_) in
+            self?.notifyObservers(percent: 0, completed: false, error: nil)
+            self?.inProgress = true
+            self?.progressValue = 0.0
+            self?.processCloudFiles()
+        }
+
+        NotificationCenter.default.addObserver(forName: NSNotification.Name.NSMetadataQueryDidUpdate, object: query, queue: query.operationQueue) { [weak self] (_) in
+            self?.processCloudFiles()
+        }
+    }
+
+    private func processCloudFiles() {
+        if query.resultCount == 0 { return }
+        var fileItem: NSMetadataItem?
+        var fileURL: URL?
+
+        for item in query.results {
```

Comment:
Are we sure that `query.results` is sorted as we expect?

Reply:
yes